File: AMF5/src/test/java/javaPlatform/ErrorHandlerTest.java
package javaPlatform;

import amf.apicontract.client.platform.AMFBaseUnitClient;
import amf.apicontract.client.platform.RAMLConfiguration;
import amf.core.client.platform.AMFResult;
import amf.core.client.platform.errorhandling.ClientErrorHandler;
import amf.core.client.platform.validation.AMFValidationResult;
import org.junit.Test;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutionException;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThrows;

public class ErrorHandlerTest {


    class UnhandledErrorHandler implements ClientErrorHandler {

        @Override
        public void report(AMFValidationResult result) {
            throw new RuntimeException(result.message());
        }

        @Override
        public List<AMFValidationResult> getResults() {
            return Collections.emptyList();
        }
    }

    @Test
    public void customErrorHandler() throws ExecutionException, InterruptedException {
        final ClientErrorHandler eh = new UnhandledErrorHandler();
        final AMFBaseUnitClient client = RAMLConfiguration.RAML10()
                .withErrorHandlerProvider(() -> eh)
                .baseUnitClient();

        final AMFResult parseResult = client.parse("file://src/test/resources/examples/resolution-error.raml").get();

        assertNotNull(parseResult.baseUnit());

        assertThrows(RuntimeException.class,
                () -> client.transform(parseResult.baseUnit()));
    }
}

File: AMF5/src/test/scala/scalaPlatform/ErrorHandlerTest.scala
package scalaPlatform

import amf.apicontract.client.scala.RAMLConfiguration
import amf.core.client.scala.errorhandling.UnhandledErrorHandler
import org.junit.runner.RunWith
import org.scalatest.flatspec.AsyncFlatSpec
import org.scalatest.matchers.should
import org.scalatestplus.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class ErrorHandlerTest extends AsyncFlatSpec with should.Matchers {

  "AMF client" should "use a custom error handler provider" in {
    val client =
      RAMLConfiguration
        .RAML10()
        .withErrorHandlerProvider(() =>
          UnhandledErrorHandler
        ) // throws an exception when an error is found
        .baseUnitClient()

    client.parse("file://src/test/resources/examples/resolution-error.raml") map { parseResult =>
      assertThrows[java.lang.Exception] {
        client.transform(parseResult.baseUnit)
      }
    }
  }
}
